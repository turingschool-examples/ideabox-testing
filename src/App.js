import React, { Component } from 'react';
import Ideas from './Ideas';
import Form from './Form';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      ideas: [],
      error: ''
    }
  }

  componentDidMount() {
    fetch('http://localhost:3001/api/v1/ideas')
      .then(response => response.json())
      .then(ideas => this.setState({ ideas }))
      .catch(error => this.setState({ error: error.message }));
  }

  addIdea = (newIdea) => {
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ ...newIdea })
    };

    fetch('http://localhost:3001/api/v1/ideas', options)
      .then(response => response.json())
      .then(response => fetch(`http://localhost:3001/api/v1/ideas/${response.id}`))
      .then(response => response.json())
      .then(newIdea => this.setState({ ideas: [...this.state.ideas, newIdea] }))
      .catch(error => this.setState({ error: error.message }))
  }

  deleteIdea = (id) => {
    const options = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    }

    fetch(`http://localhost:3001/api/v1/ideas/${id}`, options)
      .then(() => fetch('http://localhost:3001/api/v1/ideas'))
      .then(response => response.json())
      .then(ideas => this.setState({ ideas }))
      .catch(error => this.setState({ error: error.message }));
  }

  render() {
    return(
      <main className='App'>
        <h1>IdeaBox</h1>
        <Form addIdea={this.addIdea} />
        {this.state.error && <h2>{this.state.error}</h2>}
        <Ideas ideas={this.state.ideas} deleteIdea={this.deleteIdea} />
      </main>
    )
  }
}

export default App;
