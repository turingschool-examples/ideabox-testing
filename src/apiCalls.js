export const getIdeas = async () => {
  try {
    const response = await fetch('http://localhost:3001/api/v1/ideas');
    const ideas = await response.json();

    return ideas;
  } catch (error) {
    throw Error('Error fetching ideas');
  }
}

export const getSpecificIdea = async (id) => {
  try {
    const response = await fetch(`http://localhost:3001/api/v1/ideas/${id}`);
    const idea = await response.json();

    return idea;
  } catch (error) {
    throw Error('Error fetching idea');
  }
}

export const createIdea = newIdea => {
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ ...newIdea })
  };

  return fetch('http://localhost:3001/api/v1/ideas', options)
  .then(response => {
    if (!response.ok) {
      throw Error('Error creating idea')
    } else {
      return response.json()
    }
  })
}

export const removeIdea = async (id) => {
  try {
    const options = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    }

    await fetch(`http://localhost:3001/api/v1/ideas/${id}`, options);

    return;
  } catch(error) {
    throw Error('Error deleting idea');
  }
}
