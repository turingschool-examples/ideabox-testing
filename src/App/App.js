import React, { Component } from 'react';
import Form from '../Form/Form';
import Ideas from '../Ideas/Ideas'
import { getIdeas, postIdea, deleteIdea } from '../apiCalls';
import './App.css';

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      ideas: [],
      isLoading: true,
      error: ''
    };
  }

  componentDidMount() {
    getIdeas()
      .then(ideas => this.setState({ ideas, isLoading: false }))
      .catch(error => this.setState({
        isLoading: false,
        error: error.message
      })
    );
  }

  addIdea = newIdea => {
    postIdea(newIdea)
      .then(idea => this.setState({
        ideas: [...this.state.ideas, idea]
      }))
      .catch(error => this.setState({ error: error.message }))
  }

  removeIdea = id => {
    deleteIdea(id)
      .then(ideas => this.setState({ ideas }))
      .catch(error => this.setState({ error: error.message }));
  }

  render() {
    const { ideas, isLoading, error } = this.state;
    return (
      <main className="App">
        <h1>IdeaBox</h1>
        <Form addIdea={this.addIdea} />
        {isLoading && <img
          src={'https://www.gearbubble.com/assets/loader_large.gif'}
          alt={''}
        />
        }
        {error && <h2>{error}</h2>}
        <Ideas
          ideas={ideas}
          removeIdea={this.removeIdea}
        />
      </main>
    )
  }
}