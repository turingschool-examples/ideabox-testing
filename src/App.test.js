import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';
import { createIdea, getSpecificIdea, removeIdea, getIdeas } from './apiCalls';
jest.mock('./apiCalls', () => ({
  createIdea: jest.fn().mockImplementation(() => {
    return {id: 3}
  }),
  getSpecificIdea: jest.fn().mockImplementation(() => {
    return {id: 3, title: 'Sweaters for pugs', description: 'Why not?'}
  }),
  removeIdea: jest.fn(),
  getIdeas: jest.fn().mockImplementation(() => {
    return [{id: 3, title: 'Sweaters for pugs', description: 'Why not?'}]
  })
}));

describe.only('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should match snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when addIdea is called', () => {
    // wrapper = shallow(<App />, { disableLifecycleMethods: true });

    const mockIdea = {
      id: 3, title: 'Sweaters for pugs', description: 'Why not?'
    };
    const expected = [mockIdea];

    wrapper.instance().addIdea(mockIdea);

    expect(wrapper.state('ideas')).toEqual(expected);
  });

  it('should update state when deleteIdea is called', async () => {
    const expected = [{id: 3, title: 'Sweaters for pugs', description: 'Why not?'}];

    wrapper.instance().setState({ ideas: [{ id: 1, title: 'Prank Travis', description: 'Stick googly eyes on all his stuff' }, {id: 3, title: 'Sweaters for pugs', description: 'Why not?'}] });

    await wrapper.instance().deleteIdea(1);

    expect(wrapper.state('ideas')).toEqual(expected);
  });
});
