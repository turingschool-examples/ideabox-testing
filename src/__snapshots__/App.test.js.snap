// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<main
  className="App"
>
  <h1>
    IdeaBox
  </h1>
  <Form
    addIdea={[Function]}
  />
  <Ideas
    deleteIdea={[Function]}
    ideas={
      Array [
        Object {
          "description": "Why not?",
          "id": 3,
          "title": "Sweaters for pugs",
        },
      ]
    }
  />
</main>
`;
